### PR TITLE
JAMES-3440 API, Memory implementation and contract test for EmailQuer…

### PR DESCRIPTION
**Pull request by chibenwa**
…yView

### CODE REVIEW

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/projections/MemoryEmailQueryView.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            return Objects.hash(mailboxId, messageId, sentAt, receivedAt);
+        }
+    }
+
+    @Inject
+    public MemoryEmailQueryView() {
+        entries = Tables.synchronizedTable(HashBasedTable.create());
+    }
+
+    private final Table<MailboxId, MessageId, Entry> entries;
```
Comment (Arsnael) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/projections/MemoryEmailQueryView.java:95]: should be declared before the public constructor

**Comment by chibenwa** [comment]
Please note that https://github.com/apache/james-project/pull/259 contains explanations regarding table structure.

**Comment by Arsnael** [comment]
Why not separating memory and cassandra implems into 2 PRs?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryView.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    private final CassandraMessageId.Factory messageIdFactory;
+    private final CassandraAsyncExecutor executor;
+    private final PreparedStatement listMailboxContentBySentAt;
+    private final PreparedStatement listMailboxContentSinceSentAt;
+    private final PreparedStatement listMailboxContentSinceReceivedAt;
+    private final PreparedStatement insertInLookupTable;
+    private final PreparedStatement insertReceivedAt;
+    private final PreparedStatement insertSentAt;
+    private final PreparedStatement deleteLoopupRecord;
```
Comment (Arsnael) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryView.java:73]: s/deleteLoopupRecord/deleteLookupRecord

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryView.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    private final PreparedStatement deleteSentAt;
+    private final PreparedStatement deleteReceivedAt;
+    private final PreparedStatement deleteAllLoopupRecords;
```
Comment (Arsnael) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryView.java:76]: s/deleteAllLoopupRecords/deleteAllLookupRecords

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryView.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+        CassandraId cassandraId = (CassandraId) mailboxId;
+        return executor.executeRows(listMailboxContentBySentAt.bind()
+                .setUUID(MAILBOX_ID, cassandraId.asUuid())
+                .setInt(LIMIT_MARKER, limit.getLimit().get()))
+            .map(row -> messageIdFactory.of(row.getUUID(MESSAGE_ID)));
+    }
+
+    @Override
+    public Flux<MessageId> listMailboxContentSinceReceivedAt(MailboxId mailboxId, ZonedDateTime since, Limit limit) {
+
```
Comment (Arsnael) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryView.java:169]: blank line

File: server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryViewTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public MessageId messageId3() {
+        return MESSAGE_ID_3;
+    }
+
+    @Override
+    public MessageId messageId4() {
+        return MESSAGE_ID_4;
+    }
```
Comment (Arsnael) [inline server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryViewTest.java:79]: should we have some concurrency tests as well?

Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryViewTest.java:79]: What for?

Which one?

Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/projections/CassandraEmailQueryViewTest.java:79]: Each messageId is unique in a mailbox...

**Review by Arsnael — APPROVED** [review]

**Review by LanKhuat — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**